cosmos/cosmos-sdk
# feat(textual): Add client-side infra

## Description

Closes: #13747

- Add clientCtx support and CLI flags for Textual (will throw an error for now if used)
- Add keyring support for Ledger devices
  - To tell the device whether the bytes we're sending are TEXTUAL bytes or AMINO_JSON bytes, we pass a byte flag called `P2`, which can be 0 or 1. To achieve this, I needed to break some keyring APIs for this new argument, as well as add a new `LedgerPrivKeyAminoJSON` interface so that ledger priv keys have 2 `Sign` methods.
  - Let me know if you can think of better ways to achieve this.
- Add ante handler tests to make sure Textual passes sig verification.

## Manual Testing

I managed to sign on my Nano S by performing the following changes on this PR ([see as diff](https://github.com/cosmos/cosmos-sdk/compare/am/13747-textual-client...am/e2e-ledger-test?expand=1)):
- Install the new Cosmos app >=v2.35.1 of https://github.com/cosmos/ledger-cosmos
- remove [these lines](https://github.com/cosmos/cosmos-sdk/pull/14661/files#diff-1c3172807f17252df50066b921f45e4804f7cada111740b51f505a6a3c15716cR284-R289) which are CLI guards which prevent Textual
- enable SIGN_MODE_TEXTUAL in [DefaultSignModes](https://github.com/cosmos/cosmos-sdk/blob/44fbb0df9cea049d588e76bf930177d777552cf3/x/auth/tx/mode_handler.go#L12)
- Add these two lines on top of `simapp/app_v2.go` (won't be necessary after #14647)
```go
	_ "cosmossdk.io/api/cosmos/bank/v1beta1"
	_ "cosmossdk.io/api/cosmos/crypto/secp256k1"
```
---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...
